### PR TITLE
Update recommendations for ScalarDB

### PR DESCRIPTION
**Pull request by kota2and3kan**
## Description

We updated our recommendations for the production environment. So, I updated some documents based on the new recommendations.
These updates are based on the [Production checklist for ScalarDB Cluster](https://github.com/scalar-labs/scalar-kubernetes/blob/master/docs/ProductionChecklistForScalarDBCluster.md).

## Related issues and/or PRs

We discussed the new recommendation in #207.

## Changes made

* Update each document based on the production checklist (new recommendations).
* Update diagrams based on the production checklist (new recommendations).

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes.
- [x] Any remaining open issues linked to this PR are documented and up-to-date (Jira, GitHub, etc.).
- [x] Tests (unit, integration, etc.) have been added for the changes.
- [x] My changes generate no new warnings.
- [x] Any dependent changes in other PRs have been merged and published.

## Additional notes (optional)

N/A

### CODE REVIEW

File: docs/CreateAKSClusterForScalarDB.md [hunk 1]
```diff
@@ -21,32 +21,38 @@ When deploying ScalarDB Server, you must:
...
+
+If you place the worker nodes in different [availability zones](https://learn.microsoft.com/en-us/azure/availability-zones/az-overview) (AZs), you can withstand an AZ failure.
+{% endcapture %}
+
+<div class="notice--info">{{ notice--info | markdownify }}</div>
+
 ### Use 4vCPU / 8GB memory nodes for the worker node in the ScalarDB Server node pool
 
-From the perspective of commercial licenses, resources for one pod running ScalarDB Server are limited to 2vCPU / 4GB memory. In addition, we recommend deploying one ScalarDB Server pod and one Envoy pod on one worker node.
+From the perspective of commercial licenses, resources for one pod running ScalarDB Server  are limited to 2vCPU / 4GB memory. In addition, some pods other than ScalarDB Server pods exist on the worker nodes.
```
Comment (feeblefakie) [inline docs/CreateAKSClusterForScalarDB.md:38]: ```In addition, some pods other than ScalarDB Server pods exist on the worker nodes.```
sounds like some other pods **always** exist on the worker node, but the following sentence says other components **could** run on the worker node, so I felt a bit inconsistency between them.
I might miss something though.

Comment (kota2and3kan) [reply docs/CreateAKSClusterForScalarDB.md]: Ah, I see. You are right. It's not consistent.
I overlooked it.
I fixed it at a314ea5aad9d1db67fefbc1b7d6bcc9aec7515dd.

(I will update the same thing on other documents in another PR.)

**Comment by kota2and3kan** [comment]
@feeblefakie  
Thank you for your review!
I updated the documents based on your feedback.
Please take a look when you have time!

File: docs/CreateAKSClusterForScalarDB.md [hunk 2]
```diff
@@ -21,32 +21,38 @@ When deploying ScalarDB Server, you must:
 
 The following are some recommendations for deploying ScalarDB Server. These recommendations are not required, so you can choose whether or not to apply these recommendations based on your needs.
 
+### Create at least three worker nodes and three pods
```
Comment (feeblefakie) [inline docs/CreateAKSClusterForScalarDB.md:24]: It's a comment on the whole doc.
I'm wondering if this should be a deprecate document since the corresponding on for EKS is marked as `Deprecated`.
Can you clarify? (maybe it's related to the reason why we don't have updates for ScalarDB Cluster doc for AKS?)

**Review by feeblefakie — COMMENTED** [review]
Overall looking good!
Left one question for clarification.

File: docs/CreateAKSClusterForScalarDB.md (hunk 2 above)
Comment (kota2and3kan) [reply docs/CreateAKSClusterForScalarDB.md:24]: @feeblefakie  
Thank you for your question!

This question is related to the ScalarDB Cluster's listing in each marketplace.
At this time, we provide ScalarDB Cluster on only AWS Marketplace as follows.

* AWS Marketplace
  * ScalarDB Cluster (Pay As You Go)
  * [Deprecated] ScalarDB Server (BYOL)
* Azure Marketplace
  * ScalarDB Server (BYOL)

So, if we mark ScalarDB Server as deprecated on the Azure side, basically, Azure users cannot use ScalarDB in any way. This is why I don't still mark ScalarDB Server as deprecated on the Azure Marketplace (Azure document) side.

If it (Azure users cannot use ScalarDB) is acceptable, we can mark it as deprecated.
(Of course, from the technical perspective, users can use ScalarDB Server on Azure environment even if we mark it as a deprecated product.)

Also, there are almost no users on the Azure Marketplace (ScalarDB Server BYOL), so it might be no problem if we mark it as deprecated.
What do you think?

---

For your information, I will organize the Marketplace and documents as follows after we will release license based BYOL.

* AWS Marketplace
  * ScalarDB Cluster (Pay As You Go)
  * ScalarDB Cluster (license-key based BYOL)
  * [Deprecated] ScalarDB Server (BYOL)
* Azure Marketplace
  * ScalarDB Cluster (license-key based BYOL)
  * [Deprecated] ScalarDB Server (BYOL)

In this case, both users of AWS and Azure can use ScalarDB Cluster on each cloud environment.

File: docs/CreateEKSClusterForScalarDB.md [hunk 3]
```diff
@@ -19,27 +25,33 @@ When deploying ScalarDB Server, you must:
...
+{% endcapture %}
 
-With this in mind, you should use a worker node that has 4vCPU / 8GB memory resources. We recommend running only the above components on the worker node for ScalarDB Server. However, if you want to run other pods on the worker node for ScalarDB Server, you should use a worker node that has more than 4vCPU / 8GB memory.
+<div class="notice--info">{{ notice--info | markdownify }}</div>
 
-Note that you should configure resource limits based on your system's workload if the Envoy pod exceeds the above resource usage. In addition, you should consider scaling out the worker node and the ScalarDB Server pod if the ScalarDB Server pod exceeds the above resource usage and if latency is high (throughput is low) in your system.
+### Use 4vCPU / 8GB memory nodes for the worker node in the ScalarDB Server node group
 
-### Create a node group for other application pods than ScalarDB Server pods
+From the perspective of commercial licenses, resources for one pod running ScalarDB Server  are limited to 2vCPU / 4GB memory. In addition, some pods other than ScalarDB Server pods could exist on the worker nodes.
```
Comment (josh-wong) [inline docs/CreateEKSClusterForScalarDB.md:42]:
```suggestion
From the perspective of commercial licenses, resources for one pod running ScalarDB Server are limited to 2vCPU / 4GB memory. In addition, some pods other than ScalarDB Server pods could exist on the worker nodes.
```

File: docs/CreateEKSClusterForScalarDB.md [hunk 4]
```diff
@@ -19,27 +25,33 @@ When deploying ScalarDB Server, you must:
...
 
-We recommend running only ScalarDB Server pods on the worker node (node group) for ScalarDB Server. If you want to run other application pods on the same EKS cluster, you should create other node groups for your application pods.
+In other words, the following components could run on one worker node:
 
-### Create a node group for monitoring components (kube-prometheus-stack and loki-stack)
+* ScalarDB Server pod (2vCPU / 4GB)
+* Envoy proxy
+* Your application pods (if you choose to run your application's pods on the same worker node)
+* Monitoring components (if you deploy monitoring components such `kube-prometheus-stack`)
```
Comment (josh-wong) [inline docs/CreateEKSClusterForScalarDB.md:49]:
```suggestion
* Monitoring components (if you deploy monitoring components such as `kube-prometheus-stack`)
```

File: docs/CreateEKSClusterForScalarDB.md [hunk 5]
```diff
@@ -19,27 +25,33 @@ When deploying ScalarDB Server, you must:
...
+With this in mind, you should use a worker node that has at least 4vCPU / 8GB memory resources and use at least three worker nodes for availability, as mentioned in [Create at least three worker nodes and three pods](#create-at-least-three-worker-nodes-and-three-pods).
+
+However, three nodes with at least 4vCPU / 8GB memory resources per node is the minimum for a production environment. You should also consider the resources of the EKS cluster (for example, the number of worker nodes, vCPUs per node, memories per node, ScalarDB Server pods, and pods for your application), which depend on your system's workload. In addition, if you plan to scale the pods automatically by using some features like [Horizontal Pod Autoscaling (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/), you should consider the maximum number of pods on the worker node to decide on the worker node resources.
```
Comment (josh-wong) [inline docs/CreateEKSClusterForScalarDB.md:54]:
```suggestion
However, three nodes with at least 4vCPU / 8GB memory resources per node is the minimum for a production environment. You should also consider the resources of the EKS cluster (for example, the number of worker nodes, vCPUs per node, memories per node, ScalarDB Server pods, and pods for your application), which depend on your system's workload. In addition, if you plan to scale the pods automatically by using some features like [Horizontal Pod Autoscaling (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/), you should consider the maximum number of pods on the worker node when deciding the worker node resources.
```

File: docs/CreateEKSClusterForScalarDB.md [hunk 6]
```diff
@@ -51,18 +63,11 @@ In addition, if you configure Cluster Autoscaler, you should create a subnet in
...
-
-* Create three subnets in different availability zones in the VPC.
-* Create at least three worker nodes.
-
 ### Restrict connections by using some security features based on your requirements
 
 You should restrict unused connections in ScalarDB Server. To restrict unused connections, you can use some security features in AWS, like [security groups](https://docs.aws.amazon.com/vpc/latest/userguide/VPC_SecurityGroups.html) and [network access control lists](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-network-acls.html).
 
-The connections (ports) that ScalarDB Server uses by default are as follows. Note that, if you change the default listening port for ScalarDB Server in the configuration file (`database.properties`), you must allow connections by using the port that you configured.
+The connections (ports) that ScalarDB Server uses by default are as follows.
```
Comment (josh-wong) [inline docs/CreateEKSClusterForScalarDB.md:70]:
```suggestion
The connections (ports) that ScalarDB Server uses by default are as follows:
```

File: docs/CreateEKSClusterForScalarDB.md [hunk 7]
```diff
@@ -71,70 +76,18 @@ The connections (ports) that ScalarDB Server uses by default are as follows. Not
...
+
+<div class="notice--info">{{ notice--info | markdownify }}</div>
+
+{% capture notice--info %}
+**Note**
+
+You also must allow the connections that EKS uses itself. For more details about Amazon EKS security group requirements, refer to [Amazon EKS security group requirements and considerations](https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html).
+{% endcapture %}
+
+<div class="notice--info">{{ notice--info | markdownify }}</div>
```
Comment (josh-wong) [inline docs/CreateEKSClusterForScalarDB.md:93]: I've combined these two note admonitions/callouts since having them stacked might look a little awkward on the docs site.

```suggestion
{% capture notice--info %}
**Note**

- If you change the default listening port for ScalarDB Server in the configuration file (`database.properties`), you must allow connections by using the port that you configured.
- You also must allow the connections that EKS uses itself. For more details about Amazon EKS security group requirements, refer to [Amazon EKS security group requirements and considerations](https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html).
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>
```

File: docs/CreateEKSClusterForScalarDBCluster.md [hunk 8]
```diff
@@ -17,27 +17,41 @@ When deploying ScalarDB Cluster, you must:
...
+From the perspective of commercial licenses, resources for one pod running ScalarDB Cluster are limited to 2vCPU / 4GB memory. In addition, some pods other than ScalarDB Cluster pods could exist on the worker nodes.
 
-In other words, the following components run on one worker node:
+In other words, the following components could run on one worker node:
 
 * ScalarDB Cluster pod (2vCPU / 4GB)
-* Envoy proxy
+* Envoy proxy (if you use `indirect` client mode or use a programming language other than Java)
+* Your application pods (if you choose to run your application's pods on the same worker node)
+* Monitoring components (if you deploy monitoring components such `kube-prometheus-stack`)
```
Comment (josh-wong) [inline docs/CreateEKSClusterForScalarDBCluster.md:41]:
```suggestion
* Monitoring components (if you deploy monitoring components such as `kube-prometheus-stack`)
```

File: docs/CreateEKSClusterForScalarDBCluster.md [hunk 9]
```diff
@@ -49,18 +63,11 @@ In addition, if you configure Cluster Autoscaler, you should create a subnet in
...
-
-* Create three subnets in different availability zones in the VPC.
-* Create at least three worker nodes.
-
 ### Restrict connections by using some security features based on your requirements
 
 You should restrict unused connections in ScalarDB Cluster. To restrict unused connections, you can use some security features in AWS, like [security groups](https://docs.aws.amazon.com/vpc/latest/userguide/VPC_SecurityGroups.html) and [network access control lists](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-network-acls.html).
 
-The connections (ports) that ScalarDB Cluster uses by default are as follows. Note that, if you change the default listening port for ScalarDB Cluster in the configuration file (`scalardb-cluster-node.properties`), you must allow connections by using the port that you configured.
+The connections (ports) that ScalarDB Cluster uses by default are as follows.
```
Comment (josh-wong) [inline docs/CreateEKSClusterForScalarDBCluster.md:70]:
```suggestion
The connections (ports) that ScalarDB Cluster uses by default are as follows:
```

File: docs/CreateEKSClusterForScalarDBCluster.md [hunk 10]
```diff
@@ -17,27 +17,41 @@ When deploying ScalarDB Cluster, you must:
...
 
-We recommend running only pods related to ScalarDB Cluster on the worker node for ScalarDB Cluster. If you want to run monitoring pods (e.g., Prometheus, Grafana, Loki, etc.) by using [kube-prometheus-stack](./K8sMonitorGuide.md) and [loki-stack](./K8sLogCollectionGuide.md) on the same EKS cluster, you should create other node groups for monitoring pods.
+However, three nodes with at least 4vCPU / 8GB memory resources per node is the minimum for a production environment. You should also consider the resources of the EKS cluster (for example, the number of worker nodes, vCPUs per node, memories per node, ScalarDB Cluster pods, and pods for your application), which depend on your system's workload. In addition, if you plan to scale the pods automatically by using some features like [Horizontal Pod Autoscaling (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/), you should consider the maximum number of pods on the worker node to decide on the worker node resources.
```
Comment (josh-wong) [inline docs/CreateEKSClusterForScalarDBCluster.md:54]:
```suggestion
However, three nodes with at least 4vCPU / 8GB memory resources per node is the minimum for a production environment. You should also consider the resources of the EKS cluster (for example, the number of worker nodes, vCPUs per node, memories per node, ScalarDB Cluster pods, and pods for your application), which depend on your system's workload. In addition, if you plan to scale the pods automatically by using some features like [Horizontal Pod Autoscaling (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/), you should consider the maximum number of pods on the worker node when deciding the worker node resources.
```

File: docs/CreateEKSClusterForScalarDBCluster.md [hunk 11]
```diff
@@ -70,94 +77,18 @@ The connections (ports) that ScalarDB Cluster uses by default are as follows. No
...
+
+<div class="notice--info">{{ notice--info | markdownify }}</div>
+
+{% capture notice--info %}
+**Note**
+
+You also must allow the connections that EKS uses itself. For more details about Amazon EKS security group requirements, refer to [Amazon EKS security group requirements and considerations](https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html).
+{% endcapture %}
+
+<div class="notice--info">{{ notice--info | markdownify }}</div>
```
Comment (josh-wong) [inline docs/CreateEKSClusterForScalarDBCluster.md:94]:
```suggestion
{% capture notice--info %}
**Note**

- If you change the default listening port for ScalarDB Cluster in the configuration file (`scalardb-cluster-node.properties`), you must allow connections by using the port that you configured.
- You also must allow the connections that EKS uses itself. For more details about Amazon EKS security group requirements, refer to [Amazon EKS security group requirements and considerations](https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html).
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>
```

**Review by josh-wong — CHANGES_REQUESTED** [review]
I've added some minor suggestions. PTAL!

**Review by komamitsu — APPROVED** [review]
LGTM! 👍

File: conf/scalardb-custom-values.yaml [hunk 12]
```diff
@@ -17,34 +17,22 @@ envoy:
...
+                - key: app.kubernetes.io/name
+                  operator: In
+                  values:
+                    - scalardb
+                - key: app.kubernetes.io/app
+                  operator: In
+                  values:
+                    - envoy
+            topologyKey: kubernetes.io/hostname
+          weight: 50
```
[the remark this answers is not in the document]
Comment (kota2and3kan) [reply conf/scalardb-custom-values.yaml:35]: @komamitsu  
Thank you for your review and question!

Actually, this value has almost no meaning in this sample configuration.

The `weight` value is used when Kubernetes decides the priority if there are two or more affinity configurations. You can see the sample configuration that has two affinity configurations in the following official document.
https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity-weight

As the above, the `weight` has meaning when there are two or more affinity configurations, but it has almost no meaning if there is only one affinity configuration.

However, this value is the `required` parameter as follows.

```console
$ kubectl explain pod.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution
KIND:       Pod
VERSION:    v1

FIELD: preferredDuringSchedulingIgnoredDuringExecution <[]WeightedPodAffinityTerm>

DESCRIPTION:
    The scheduler will prefer to schedule pods to nodes that satisfy the
    anti-affinity expressions specified by this field, but it may choose a node
    that violates one or more of the expressions. The node that is most
    preferred is the one with the greatest sum of weights, i.e. for each node
    that meets all of the scheduling requirements (resource request,
    requiredDuringScheduling anti-affinity expressions, etc.), compute a sum by
    iterating through the elements of this field and adding "weight" to the sum
    if the node has pods which matches the corresponding podAffinityTerm; the
    node(s) with the highest sum are the most preferred.
    The weights of all of the matched WeightedPodAffinityTerm fields are added
    per-node to find the most preferred node(s)

FIELDS:
  podAffinityTerm       <PodAffinityTerm> -required-
    Required. A pod affinity term, associated with the corresponding weight.

  weight        <integer> -required-
    weight associated with matching the corresponding podAffinityTerm, in the
    range 1-100.
```

FYI: You also can see the same information in the following API reference.
https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#PodAntiAffinity

If we don't set `wight`, we will face the following error and you cannot deploy pods.
```console
$ helm install scalardb-cluster scalar-labs/scalardb-cluster -f ./scalardb-cluster.yaml
Error: INSTALLATION FAILED: Deployment.apps "scalardb-cluster-node" is invalid: spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].weight: Invalid value: 0: must be in the range 1-100
```

So, we have to set this value even if there is only one affinity configuration. But, as I mentioned, it has almost no meaning.

Therefore, I just set the middle value `50` (this parameter can accept the range `1-100`). This is why the value is `50`. Actually, there is no strong reason for this value.

[the remark this answers is not in the document]
Comment (komamitsu) [reply conf/scalardb-custom-values.yaml:35]: @kota2and3kan Thanks for the explanation! Understood.

**Comment by kota2and3kan** [comment]
@josh-wong  
Thank you for your review!
I applied your suggestions (including suggestions in another PR #213).
Please take a look again when you have time!

**Review by brfrn169 — APPROVED** [review]
LGTM! Thank you!

File: docs/CreateAKSClusterForScalarDB.md [hunk 13]
```diff
@@ -21,32 +21,38 @@ When deploying ScalarDB Server, you must:
...
+
+<div class="notice--info">{{ notice--info | markdownify }}</div>
+
 ### Use 4vCPU / 8GB memory nodes for the worker node in the ScalarDB Server node pool
 
-From the perspective of commercial licenses, resources for one pod running ScalarDB Server are limited to 2vCPU / 4GB memory. In addition, we recommend deploying one ScalarDB Server pod and one Envoy pod on one worker node.
+From the perspective of commercial licenses, resources for one pod running ScalarDB Server are limited to 2vCPU / 4GB memory. In addition, some pods other than ScalarDB Server pods could exist on the worker nodes.
 
-In other words, the following components run on one worker node:
+In other words, the following components could run on one worker node:
```
Comment (josh-wong) [inline docs/CreateAKSClusterForScalarDB.md:40]: Does the wording for this depend on how we word a similar sentence in a different PR?: [related comment](https://github.com/scalar-labs/scalar-kubernetes/pull/213/files#r1382880277)

File: docs/CreateEKSClusterForScalarDB.md [hunk 14]
```diff
@@ -19,27 +25,33 @@ When deploying ScalarDB Server, you must:
...
-With this in mind, you should use a worker node that has 4vCPU / 8GB memory resources. We recommend running only the above components on the worker node for ScalarDB Server. However, if you want to run other pods on the worker node for ScalarDB Server, you should use a worker node that has more than 4vCPU / 8GB memory.
+<div class="notice--info">{{ notice--info | markdownify }}</div>
+
+### Use 4vCPU / 8GB memory nodes for the worker node in the ScalarDB Server node group
 
-Note that you should configure resource limits based on your system's workload if the Envoy pod exceeds the above resource usage. In addition, you should consider scaling out the worker node and the ScalarDB Server pod if the ScalarDB Server pod exceeds the above resource usage and if latency is high (throughput is low) in your system.
+From the perspective of commercial licenses, resources for one pod running ScalarDB Server are limited to 2vCPU / 4GB memory. In addition, some pods other than ScalarDB Server pods could exist on the worker nodes.
 
-### Create a node group for other application pods than ScalarDB Server pods
+In other words, the following components could run on one worker node:
```
Comment (josh-wong) [inline docs/CreateEKSClusterForScalarDB.md:44]: Does the wording for this depend on how we word a similar sentence in a different PR?: [related comment](https://github.com/scalar-labs/scalar-kubernetes/pull/213/files#r1382880277)

File: docs/CreateEKSClusterForScalarDBCluster.md [hunk 15]
```diff
@@ -17,27 +17,41 @@ When deploying ScalarDB Cluster, you must:
...
+
 ### Use 4vCPU / 8GB memory nodes for the worker node in the ScalarDB Cluster node group
 
-From the perspective of commercial licenses, resources for one pod running ScalarDB Cluster are limited to 2vCPU / 4GB memory. In addition, we recommend deploying one ScalarDB Cluster pod and one Envoy pod on one worker node. Note that you do not need to deploy an Envoy pod when using `direct-kubernetes` mode.
+From the perspective of commercial licenses, resources for one pod running ScalarDB Cluster are limited to 2vCPU / 4GB memory. In addition, some pods other than ScalarDB Cluster pods could exist on the worker nodes.
 
-In other words, the following components run on one worker node:
+In other words, the following components could run on one worker node:
```
Comment (josh-wong) [inline docs/CreateEKSClusterForScalarDBCluster.md:36]: Does the wording for this depend on how we word a similar sentence in a different PR?: [related comment](https://github.com/scalar-labs/scalar-kubernetes/pull/213/files#r1382880277)

**Review by josh-wong — CHANGES_REQUESTED** [review]
I left a couple of comments that are related to another PR. PTAL!

File: docs/CreateAKSClusterForScalarDB.md (hunk 2 above)
Comment (kota2and3kan) [reply docs/CreateAKSClusterForScalarDB.md:24]: @feeblefakie  
As we discussed, now we are preparing to provide ScalarDB Cluster on Azure Marketplace.
So, I will mark `ScalarDB Server` as `Deprecated` in all documents in another PR.

Comment (feeblefakie) [reply docs/CreateAKSClusterForScalarDB.md:24]: @kota2and3kan Got it. Thank you!

**Review by feeblefakie — APPROVED** [review]
LGTM! Thank you!

File: docs/CreateAKSClusterForScalarDB.md (hunk 13 above)
Comment (kota2and3kan) [reply docs/CreateAKSClusterForScalarDB.md:40]: Yes! I will update this sentence after the discussion in another PR is finished!

Comment (kota2and3kan) [reply docs/CreateAKSClusterForScalarDB.md]: Fixed in ba8937f6931436217f7b57c9b0433cabcf32a8cc.

**Comment by kota2and3kan** [comment]
@josh-wong  
Thank you for your review!
I applied some suggestions in PR #213 to this PR.
Please take a look when you have time!
